Further Refinements For #2737 ("Modernize update alert")

noah-nuebling opened the pull request:
Slight refinements on top of #2737 ("Modernize update alert") based on the [mockups I posted there](https://github.com/sparkle-project/Sparkle/pull/2737#issuecomment-3201522112) shortly after the pull request was merged.


## Screenshots

### Overview

I did my best to create great visual harmony and balance everywhere, no matter which elements of the SUUpdateAlert are shown or hidden:

#### Tahoe

<details>
<summary> Click here to reveal! </summary>

<img width="690" height="577" alt="Screenshot 2025-08-21 at 9 48 15 PM" src="https://github.com/user-attachments/assets/d640e1ec-f37f-4c58-93fe-1bb2a517387c" />
<img width="690" height="577" alt="Screenshot 2025-08-21 at 9 48 25 PM" src="https://github.com/user-attachments/assets/121d5ae4-55dd-4953-afcd-73a7a39a0f5a" />
<img width="690" height="286" alt="Screenshot 2025-08-21 at 9 48 54 PM" src="https://github.com/user-attachments/assets/5541720d-b9fb-4a9d-af99-28b44183ae6e" />
<img width="690" height="308" alt="Screenshot 2025-08-21 at 9 49 10 PM" src="https://github.com/user-attachments/assets/aaab000a-7023-46a2-ba06-01dcad23e66d" />
</details>

#### Sequoia

<details>
<summary> Click here to reveal! </summary>

<img width="678" height="580" alt="Screenshot 2025-08-21 at 21 45 29" src="https://github.com/user-attachments/assets/e37d17c7-2733-421c-92bc-aa3ccfbc54fa" />
<img width="678" height="604" alt="Screenshot 2025-08-21 at 21 45 39" src="https://github.com/user-attachments/assets/439aaa41-6302-4347-87fd-242ffd89011e" />
<img width="678" height="271" alt="Screenshot 2025-08-21 at 21 46 46" src="https://github.com/user-attachments/assets/4f8718ac-77bb-44cb-b2ea-b173e82beb2f" />
<img width="678" height="293" alt="Screenshot 2025-08-21 at 21 46 56" src="https://github.com/user-attachments/assets/463c4694-d26d-4a4f-b912-7daa5d364335" />
</details>

### Darkmode

Darkmode has the same colors as before – still looks good.

<details>
<summary> Click here to reveal! </summary>

Tahoe:
<img width="690" height="577" alt="Screenshot 2025-08-21 at 9 50 27 PM" src="https://github.com/user-attachments/assets/40ea9e76-b731-4228-885d-95757e7d6d3d" />

Sequoia:
<img width="678" height="604" alt="Screenshot 2025-08-21 at 21 47 25" src="https://github.com/user-attachments/assets/6486f4f2-eed7-48f7-9ecb-13f316198254" />

</details>

### Variations

The code contains a couple of `if ((0))` and `if ((1))` guards for alternate stylings, because I wasn't 100% sure about some details. I can delete those of course before this is merged. You can find all these by grepping for `if ((0))` and `if ((1))`.

#### Variation: Transparent titlebar

This is closer to what's currently in #2737. I personally prefer the designs with a visible titlebar, but the code supports pixel-perfect layouts for both styles and you can easily flip back and forth.

<details>
<summary> <b> Click here to reveal </b> </summary>

Tahoe:
<img width="690" height="577" alt="Screenshot 2025-08-21 at 9 54 18 PM" src="https://github.com/user-attachments/assets/8a0e80b6-38c8-4d63-b199-8d3f244ed346" />

Sequoia:
<img width="678" height="604" alt="Screenshot 2025-08-21 at 21 53 00" src="https://github.com/user-attachments/assets/c4f6a21a-8722-4769-8099-84df0700b029" />

</details>

#### Variation: Liquid Glass

Here's another fun one – liquid glass! 

I'm not sure it's supposed to be used this way, but it looks kinda cool, and it does create some color contrast between the update notes and the window background, which is otherwise missing on Tahoe. (Cause everything is white)

I haven't spent much time trying to refine this, so if there's any interest I can put more work into that.

<details>
<summary> <b> Click here to reveal </b> </summary>

<img width="957" height="636" alt="Liquid Glass" src="https://github.com/user-attachments/assets/4e8a8966-ec94-425f-92b7-c046b04078b2" />

</details>

## Code Questions

There are a few questionable things about the code currently, which may need to be changed before merging:

- Lots of comments
    - I tend to write lots of comments. I left them in cause I thought they might help with the review, but they should probably be removed or cleaned up before merging. Just let me know and I'll do that.
- 'Spacer' views
    - For greater flexibility with the spacing I removed the 'native' NSStackView spacing and instead inserted 'spacer' elemements which are then hidden or resized by the code. I think this makes it quite transparent what is happening, but it is a little ugly in Interface Builder. If it is prefererable, these spacer views could probably also be replaced with NSLayoutConstraints or `-[NSStackView setCustomSpacing:afterView:].
- I used the commits as my personal notepad
    - They probably need to be squashed or something. I'm not sure how that works. 
- If I did anything else stupid, I'm sorry, please be patient with me! It is my first time contributing code to a collaborative open source project, and I'm doing my best to be helpful.

## Other considerations and TODOs

- If you accept this, perhaps there should also be slight tweaks to SUStatus.xib or other UIs to bring them in line.

## Misc Checklist

- [ ] My change requires a documentation update on [Sparkle's website repository](https://github.com/sparkle-project/sparkle-project.github.io)
- [ ] My change requires changes to generate_appcast, generate_keys, or sign_update

## Testing

I tested and verified my change by using one or multiple of these methods:

- [x] Sparkle Test App
- [ ] Unit Tests
- [ ] My own app
- [ ] Other (please specify)

(Describe all the cases that were tested)

macOS version tested: macOS 15.5 (Sequoia) and macOS 26.0 Beta 7 (Tahoe)

noah-nuebling commented:
I can also go through the xib file and undo all the changes that Xcode does automatically whenever you open the file. That might minimize conflicts or make things easier to review.

noah-nuebling commented:
I forgot to test the layout when the **'question' text has 2 lines**, but it works quite well:

Tahoe:
<img width="690" height="577" alt="Bildschirmfoto 2025-08-23 um 9 55 56 AM" src="https://github.com/user-attachments/assets/ac29e30d-5c81-4a28-b2b1-73640b14304e" />

Sequoia:
<img width="678" height="604" alt="Bildschirmfoto 2025-08-23 um 10 01 44" src="https://github.com/user-attachments/assets/fec47bb2-0540-4f1d-96a5-c4f465b0e67d" />

---

I also tested on **Mojave / Catalina** and found some issues with the initial pull request. But I've fixed those, and now it looks like this!

<img width="732" height="589" alt="Screenshot 2025-08-23 at 11 44 07" src="https://github.com/user-attachments/assets/74ec9298-6707-49f2-8472-d7876351bb66" />

zorgiepoo commented:
Thanks for opening this.

I think overall the UI changes here are very subtle but the code changes are very non-trivial / custom. Although personally I think I prefer the UI in 2.x (I think I like the transparent titlebar). At this point I think I'd expect a single small change (both UI and code delta change) with good justification.

zorgiepoo commented:
> (I'm not sure how to test the SUUpdatePermissionPrompt.xib. I only know it exists because it showed up for me once randomly in the Sparkle Test App.)

Delete the test app user defaults. On second launch it will show permission prompt. This works when apps do not override [SUEnableAutomaticChecks](https://sparkle-project.org/documentation/customization/) in their Info.plist

noah-nuebling commented:
Ohh man I just spent the evening writing up a detailed post and then I dragged a .png onto the comment field and it tried to open the image in Safari and failed, and then I went back  and the entire comment text was deleted. :'(

I'm sorry I don't have the patience to rewrite something good, so I'll just post all the images I had made for the previous post and write some sloppy stuff around that

I made this image to show how to activate the 'transparent titlebar' style which you preferred. (The code supports that)

<img width="704" height="174" alt="Screenshot 2025-08-26 at 9 57 04 PM" src="https://github.com/user-attachments/assets/9e6b1c5d-4bd3-449b-afe1-f07e2b9a7be8" />
 it.

Then I made some images outlining the 'problems' with the current visuals that the pull request attempts to improve on:

<img width="1087" height="774" alt="Screenshot 2025-08-27 at 12 46 45 AM" src="https://github.com/user-attachments/assets/c72177a3-baca-4152-924f-632a65066107" />

<img width="1144" height="730" alt="Screenshot 2025-08-26 at 10 16 40 PM" src="https://github.com/user-attachments/assets/0ec7e8f3-a8e5-43ab-8a4f-59fb3b7b91eb" />

Here is this pull request by comparison: 

<img width="896" height="755" alt="Screenshot 2025-08-27 at 12 09 21 AM" src="https://github.com/user-attachments/assets/d9cea092-bc12-44db-ad9d-3067bc0363a4" />

Here with the margins painted in – same sized margins have the same color:

<img width="931" height="785" alt="Screenshot 2025-08-26 at 10 22 47 PM" src="https://github.com/user-attachments/assets/3c1421a5-d015-44ea-8120-f4e0cc79c7a4" />

I then wrote about how I think your eyes can't pick up pixel-differences per-se (which I think you alluded to and I agree with) but your eyes are very good at comparing things, and the new margins create lots of 'symmetries' – things are centered between other things and close-by margins are often equal. I said that I thought that this made things look more beautiful and simple and professional and stuff.

Here are some more 'symmetries' – between the text and the app icon:

The text sits nicely between the rounded corners on Tahoe:
<img width="200" alt="Screenshot 2025-08-26 at 11 18 39 PM" src="https://github.com/user-attachments/assets/e7a46937-9c87-4e7c-bbcd-7952220c6cd9" />

On Sequoia, the text lines up with the little arrow heads – I didn't plan for this and it might be an accident, but Id assume that the icon follows Apple's grid and other icons would line up in similar ways and stuff:
<img width="400" alt="Screenshot 2025-08-26 at 11 14 54 PM" src="https://github.com/user-attachments/assets/ceda7130-e392-4f05-9cff-52a2cab45307" />

I also made a case for how these symmetries show up in any situation, no matter which elements are shown or hidden or what macOS version you're using. 

E.g here's Sequoia with a transparent title bar:
Note that when the titlebar is transparent, the code adjusts the top margin so that the app icon is centered between the traffic lights and the releaseNotesBox. This works on all macOS versions that Sparkle supports (I think – I tested 10.14 and up)

<img width="921" height="770" alt="Screenshot 2025-08-26 at 11 58 19 PM" src="https://github.com/user-attachments/assets/d37dd729-7beb-4571-a421-a60f9887dbdd" />

Here are margins for when checkbox is hidden on Tahoe:

<img width="300" alt="Screenshot 2025-08-27 at 12 00 25 AM" src="https://github.com/user-attachments/assets/f9ff89b1-3e00-4448-b17c-cccd13ecbbaa" />

Margins when releaseNotesBox is hidden on Tahoe:

<img width="1000" alt="Screenshot 2025-08-27 at 12 00 39 AM" src="https://github.com/user-attachments/assets/1d7dbfcf-6023-4728-b05f-4851dee71580" />

---

I then I tried to make a case for the code:

Basically my point was that, when you strip away the if-else cases which I just left in fyi or so you can decide what you like better, the added code is just this:

<img width="1068" height="636" alt="Screenshot 2025-08-27 at 12 10 29 AM" src="https://github.com/user-attachments/assets/8413f14f-7617-4d8b-9290-db5a53376c46" />

Which is mostly just changing NSLayoutConstraint sizes on Tahoe – it's really not that complicated, it's just maybe a bit verbose because NSLayoutConstraint API.

Beyond this, there are two 'architectural changes' to the UI:

1. Change: 

The icon has been wrapped in a wrapper view which 'deletes' the margins contained in the image from the layout, by applying negative margins. This allows us to visually align the app icon with the other elements on screen very easily. It also allows us to resize the the app icon without affecting all the margins. This makes it much simpler to make the app icon bigger on Tahoe – which I think is good because otherwise the app icon looks too small on Tahoe – especially in contrast to the big buttons (But all the controls and the title bar have gotten bigger in Tahoe.) 

-> So I think this change actually allows for simplification of the layout logic and also easier iteration on the design in the future. (Because you can now very easily change the size without affecting the margins.)

2. Change:

Using 'spacer' views instead of `-[NSStackview setCustomSpacing:afterView:]` (and wrapper views with margins.)

This is almost the same thing. Spacer views can do the same things as -[setCustomSpacing:afterView:] but are more flexible and powerful, since:
1. They allow you to specify different spacing between different views in the stack right in interface builder.
2. You can use a spacer view to create a space at the *very top* of the stack view, which isn't possible with setCustomSpacing:afterView:
3. You can stack multiple spacer on top of each other and then hide one of them programmatically 

This simplifies the code in our case, because we can define the spacing at the very top of the window edge in the same way as any where else. We used this to define the `_spacer_underneath_trafficLights`, which we configure to line up with the bottom edge of the traffic-lights exactly, and then we put another spacer directly below which defines how far our content should be *away from the traffic lights.*

Also the checkbox previously had a wrapper view with a margin (Which, IIRC, stacked up with the [setCustomSpacing:afterView:] spacing, which is a bit hacky and confusing.). 

I changed it to a spacer to be consistent.

All of this would be a bit more hacky with [setCustomSpacing:afterView:] and wrapper views that contain the margins – but at the end of the day it's almost the same thing. If you want me to revert it back to match more closely with how it was I can do that.

Anyways, the code is only affected by all this in two tiny places:

<img width="1391" height="277" alt="Screenshot 2025-08-27 at 12 31 31 AM" src="https://github.com/user-attachments/assets/ef49f901-0e6d-48f3-9ac5-eb7d9e3fe4f2" />

I'd say this simplifies things. We removed the '.superview' thing which was a bit hacky. And we removed the SUUpdateAlertGroupElementSpacing stuff which is also slightly hacky.
Now the logic is just "If we wanna remove something from the stack, we just hide it and also hide the spacer below it". 
And all the spacers are set up so the layout looks perfect after you do that.

noah-nuebling commented:
> Delete the test app user defaults. On second launch it will show permission prompt. This works when apps do not override [SUEnableAutomaticChecks](https://sparkle-project.org/documentation/customization/) in their Info.plist

In my original post I also said thanks for that. Thanks! 

I also said that I understand if the changes are out of scope or don't align with your vision for the project. But since you talked about wanting a good justification for any changes I tried to make a more detailed case for why I think these changes are good.

Another thing I said was that I think the visual changes are small but I think people do pick up on these visual symmetries and balanced-ness and will perceive the UI as a bit more simple, beautiful and polished, and thoughtfully crafted, and that they will appreciate that.

But again – the changes are pretty small, and I think that if the UI ships as it is in 2.x it's already much more beautiful than before!

I just tried to improve and polish things to the best of my abilities.

zorgiepoo commented:
There is a lot of information to process here. I'm trying to stabilize 2.8 so I want to take smaller changes with low risk (if possible).

As for the leading padding on the icon images, I've noted before that [seems to](https://github.com/sparkle-project/Sparkle/pull/2737#issuecomment-3043478684) be [intentional](https://github.com/sparkle-project/Sparkle/pull/2737#issuecomment-3060614156), depending on the image I guess? I have seen that in NSAlerts and in notifications too with the same icons too. Even though it is questionable.

I'll see if I can come back to this thread and the details written here. I've other important things to work on meanwhile.

noah-nuebling commented:
> I'm trying to stabilize 2.8 so I want to take smaller changes with low risk (if possible).

That sounds reasonable. I understand. Thanks for looking into this.

> As for the leading padding on the icon images, I've noted before that seems to be intentional, depending on the image I guess? I have seen that in NSAlerts and in notifications too with the same icons too. Even though it is questionable.

That's an interesting point. One argument for aligning the icon with other things here is that the icon is like an "integrated" part of the layout for us while for standard NSAlert it feels more like it sits on top of the main layout separately.

In that notification you linked to I cannot clearly see where things are aligned. I haven't looked at it closely, yet, but that's interesting.

As for this comment you linked to:

> App icons always have additional padding I believe. Assuming the additional padding is 5 (not sure it is), the margin is still 20.

The padding is 5 if the NSImage size is 60, which it is on NSAlert - if the image is larger, then the padding increases. The code changes in this PR make the two things independent.

Good luck with the 2.8 update!

zorgiepoo commented:
Not taking changes right now and I don't think I can in the form of this PR. The icon alignment is tempting but I see app icons insetted in so many places. I believe the padding is baked into the images themselves, so it doesn't feel right to hardcode/assume a constant padding to negate by.
